### PR TITLE
opt(iterator): Optimise the iterator and introduce range iterators

### PR DESCRIPTION
**Pull request by ahsanbarkati**
This PR improves the iterator by doing in-place iteration. It also introduces range iterators.

This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/sroar/19)

### CODE REVIEW

**Review by manishrjain — APPROVED** [review]
Looks good. But, the code is hard to follow. Add comments / improve the variable names, perhaps.

I'll review it once again after that.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/sroar/19#-:-MhF2wQ5DZl4RSeIHb4k:b-rkhuad)* status: all files reviewed, 1 unresolved discussion (waiting on @ahsanbarkati)

___
*[iterator.go, line 71 at r1](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhEqn9j674K4bx7wQoP:-MhEqn9j674K4bx7wQoQ:b39qoal) ([raw file](https://github.com/dgraph-io/sroar/blob/de6cc6680eecfcd7a542c7e39a509dea4227d0f1/iterator.go#L71)):*
> ```Go
> 	}
> 	key := it.keys[2*it.kidx]
> 	off := it.keys[2*it.kidx+1]
> ```

Add comments. This code isn't the easiest to follow.

**Review by manishrjain — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/sroar/19)* status: 1 of 2 files reviewed, 5 unresolved discussions (waiting on @ahsanbarkati and @manishrjain)

___
*[iterator.go, line 31 at r2](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhFC0q-Ebxm-MKQZdnN:-MhFC0q-Ebxm-MKQZdnO:b-4w7t1w) ([raw file](https://github.com/dgraph-io/sroar/blob/256bd4584f8526d7f6c6d55a39afeec45025d5f9/iterator.go#L31)):*
> ```Go
> 	cidx int
> 
> 	bmHiIdx int // the index of current uint16 in the bitmap container, which is a list of uint16s
> ```

why not bitmapIdx? What does Hi imply? Is there a corresponding low index?
___
*[iterator.go, line 73 at r2](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhFBWvU9k1xeyMPp4uZ:-MhFBWvVC0uFqemG64Ug:b-uwp6um) ([raw file](https://github.com/dgraph-io/sroar/blob/256bd4584f8526d7f6c6d55a39afeec45025d5f9/iterator.go#L73)):*
> ```Go
> 	// it.keys is a list of the form [key1, offset1, key2, offset2 ... ]
> 	// the kidx'th key will be present at index 2*it.kidx and the offset corresponding to it will be
> 	// ont index after after it.
> ```

Can you fix this? `ont index after after it`

Also, why not just increment it.kidx by 2? Then, you can say
```key := it.keys[it.keyIdx]; off := it.keys[it.keyIdx+1]```

Easier to read.
___
*[iterator.go, line 85 at r2](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhFBuNp0cOdgoeSkDPQ:-MhFBuNp0cOdgoeSkDPR:b-ptoago) ([raw file](https://github.com/dgraph-io/sroar/blob/256bd4584f8526d7f6c6d55a39afeec45025d5f9/iterator.go#L85)):*
> ```Go
> 			return 0
> 		}
> 		it.kidx++
> ```

it.keyIdx += 2
___
*[iterator.go, line 103 at r2](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhFBxN98e6vBmh1T5Gc:-MhFBxN98e6vBmh1T5Gd:bz29n7w) ([raw file](https://github.com/dgraph-io/sroar/blob/256bd4584f8526d7f6c6d55a39afeec45025d5f9/iterator.go#L103)):*
> ```Go
> 		// A bitmap container is an array of uint16s.
> 		// If the container is bitmap, go to the index which has a non-zero value.
> 		for it.bitset == 0 && it.bmHiIdx+1 < len(cont[startIdx:]) {
> ```

bmHiIdx

**Review by ahsanbarkati — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/sroar/19)* status: 1 of 2 files reviewed, 5 unresolved discussions (waiting on @manishrjain)

___
*[iterator.go, line 71 at r1](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhEqn9j674K4bx7wQoP:-MhFI9gIAp3QtwLGanJY:b-896fix) ([raw file](https://github.com/dgraph-io/sroar/blob/de6cc6680eecfcd7a542c7e39a509dea4227d0f1/iterator.go#L71)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Add comments. This code isn't the easiest to follow.
</blockquote></details>

Done.
___
*[iterator.go, line 31 at r2](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhFC0q-Ebxm-MKQZdnN:-MhFI6Ho2YJ6R7smlfYZ:b-896fix) ([raw file](https://github.com/dgraph-io/sroar/blob/256bd4584f8526d7f6c6d55a39afeec45025d5f9/iterator.go#L31)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

why not bitmapIdx? What does Hi imply? Is there a corresponding low index?
</blockquote></details>

Done.
___
*[iterator.go, line 73 at r2](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhFBWvU9k1xeyMPp4uZ:-MhFI9E8DbovmH2XW9To:b-896fix) ([raw file](https://github.com/dgraph-io/sroar/blob/256bd4584f8526d7f6c6d55a39afeec45025d5f9/iterator.go#L73)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Can you fix this? `ont index after after it`

Also, why not just increment it.kidx by 2? Then, you can say
```key := it.keys[it.keyIdx]; off := it.keys[it.keyIdx+1]```

Easier to read.
</blockquote></details>

Done.
___
*[iterator.go, line 85 at r2](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhFBuNp0cOdgoeSkDPQ:-MhFI9vF64B3ozjzwfOq:b-896fix) ([raw file](https://github.com/dgraph-io/sroar/blob/256bd4584f8526d7f6c6d55a39afeec45025d5f9/iterator.go#L85)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

it.keyIdx += 2
</blockquote></details>

Done.
___
*[iterator.go, line 103 at r2](https://reviewable.io/reviews/dgraph-io/sroar/19#-MhFBxN98e6vBmh1T5Gc:-MhFIANa3PiFmL_ghX3b:b-896fix) ([raw file](https://github.com/dgraph-io/sroar/blob/256bd4584f8526d7f6c6d55a39afeec45025d5f9/iterator.go#L103)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

bmHiIdx
</blockquote></details>

Done.

**Review by manishrjain — APPROVED** [review]
:lgtm: approved. Feel free to merge.

> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/sroar/19#-:-MhFOmVO-d_aewEG3kgr:b-bitbg7)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ahsanbarkati)

**Comment by decentral1se** [comment]
@ahsanbarkati Sorry for resurrecting this but I'm doing upgrade work atm and am struggling to understand how to implement a reverse iterator on the latest release. Can I use the new range iterator for this purpose? Or do I need to do a `ToArray()` and then reverse from there? Thanks in advance!

If handier, I have opened up https://discuss.dgraph.io/t/how-to-implement-a-reverse-roaring-bitmap-iterator-sroar-bitmap-newreverseiterator-went-away/17901 :+1:
